Switch from swiftmailer to mailer

Zowac opened the pull request:
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | yes
| Fixed issues | -

Reduces the vendor list by replacing swiftmailer by symfony/mailer which has a better, moder and well maintained API.

Zowac committed:
Refactor:
- Changing SwiftMailer with Mailer

Zowac committed:
Adding symfony/google-mailer

Zowac committed:
Changing definition 'swiftmailer.mailer.default_transport' with 'mailer.default_transport'

Zowac committed:
Refacto:
- Using Mailer instead of SwiftMailer in Swift_Mailer
- Renaming Swift_Mailer in SymfonyMailer

Zowac committed:
Refacto:
- Removing tests from SymfonyMailer (ex SwiftMailer), no use found
- updating composer.json & api.md

Zowac committed:
Remove unnecessary code & fix InvalidArgumentException in SymfonyMailer

Zowac committed:
Add tests

chalasr commented on main/core/Library/Mailing/Client/SymfonyMailer.php:
```diff
@@ -0,0 +1,40 @@
+            $this->mailer->send($message);
+
+            return true;
+        } catch (\Exception $e) {
```
Only exceptions from the symfony/mailer component need to be caught (https://github.com/symfony/symfony/blob/5.x/src/Symfony/Component/Mailer/Exception/ExceptionInterface.php)

chalasr replied:
Also I suggest to log that an error occurred while trying to send the mail, with the exception passed in the log context.

Zowac replied:
Done, thx :)

chalasr commented on main/core/Library/Mailing/Client/SymfonyMailer.php:
```diff
@@ -0,0 +1,40 @@
+        foreach ($message->getAttribute('attachments') as $attachment) {
+            $email->attachFromPath(
+                $attachment['path']
+            );
```
nitpick: should be on one line :)

Zowac replied:
Done, thx :)

chalasr commented on main/core/Library/Mailing/TransportFactory.php:
```diff
@@ -12,72 +12,73 @@
 
 use Claroline\CoreBundle\Library\Configuration\PlatformConfigurationHandler;
+use Monolog\Logger;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
```
```suggestion
use Symfony\Contracts\EventDispatcher\EventDispatcherInterface;
```

chalasr commented on main/core/Library/Mailing/TransportFactory.php:
```diff
@@ -12,72 +12,73 @@
 namespace Claroline\CoreBundle\Library\Mailing;
 
 use Claroline\CoreBundle\Library\Configuration\PlatformConfigurationHandler;
+use Monolog\Logger;
```
Better use high-level abstractions.

```suggestion
use Psr\Log\LoggerInterface;
```

chalasr commented on main/core/Library/Mailing/TransportFactory.php:
```diff
@@ -12,72 +12,73 @@
 class TransportFactory
 {
+    private const COMMAND = '/usr/sbin/sendmail -bs';
+    private const TIME_OUT = 30;
```
in one word: `TIMEOUT` :)

chalasr commented on main/core/Library/Mailing/TransportFactory.php:
```diff
@@ -12,72 +12,73 @@
-        return $this->getBaseSmtpTransport();
-    }
+        // Default smtp
+        $encryption = null;
```
```suggestion
        $encryption = null; // Lets the transport choose the best value based on the platform.
```

chalasr commented on main/core/Library/Mailing/TransportFactory.php:
```diff
@@ -12,72 +12,73 @@
-    }
+        // Default smtp
+        $encryption = null;
+        if (!empty($this->configHandler->getParameter('mailer_encryption')) && 'none' !== $this->configHandler->getParameter('mailer_encryption')) {
```
As briefly discussed, `none` should mean "no encryption"  (i.e. `false`). The check should be updated accordingly

chalasr commented on main/core/Tests/Unit/Library/Mailing/Client/MailerTest.php:
```diff
@@ -0,0 +1,65 @@
+use Symfony\Component\Mailer\MailerInterface;
+use Symfony\Component\Mime\Email;
+
+class MailerTest extends MockeryTestCase
```
```suggestion
class SymfonyMailerTest extends MockeryTestCase
```

chalasr commented on main/core/Tests/Unit/Library/Mailing/Client/MailerTest.php:
```diff
@@ -0,0 +1,65 @@
+
+class MailerTest extends MockeryTestCase
+{
+    public function testEmailWithMessageData()
```
This test does not bring much value as it is more about testing the `Email` class which is already tested on symfony side. I suggest to remove it

chalasr commented on main/core/Tests/Unit/Library/Mailing/Client/MailerTest.php:
```diff
@@ -0,0 +1,65 @@
+        $this->assertTrue($symfonyMailer->send($message));
+    }
+
+    public function getMessageData()
```
The body of this method can be directly inlined in `testSend()` above, given it is the only relevant test

chalasr commented on main/core/DependencyInjection/Compiler/DynamicConfigPass.php:
```diff
@@ -34,8 +34,8 @@ public function process(ContainerBuilder $container)
             new Reference('Claroline\CoreBundle\Library\Mailing\TransportFactory'),
```
`Swift_Transport` does not exist anymore (line 32) :) 
The class should be `TransportInterface` from symfony/mailer (because it is a factory, we can use a broader type, no need to set concrete class).

Zowac replied:
Done, thx :)

chalasr commented on main/core/DependencyInjection/Compiler/DynamicConfigPass.php:
```diff
@@ -34,8 +34,8 @@ public function process(ContainerBuilder $container)
         );
-        $container->removeDefinition('swiftmailer.mailer.default.transport');
-        $container->setDefinition('swiftmailer.mailer.default.transport', $transport);
+        $container->removeDefinition('mailer.default_transport');
```
That `mailer.default_transport` should be configured somewhere :)
What is missing here is to replace the suggested config for `swiftmailer` by a minimal config for `mailer` (which will make the framework create the `@mailer` and `@mailer.default_transport` services).

chalasr reviewed (changes requested):
Great work so far!

Zowac committed:
Fix reviews :
- Catch exceptions from Symfony/mailer
- Fixed somes use
- Updated the checking encryption in TransportFactory
- Renamed MailerTest into SymfonyMailerTest

chalasr commented on main/core/DependencyInjection/Compiler/DynamicConfigPass.php:
```diff
@@ -29,7 +29,7 @@ public function process(ContainerBuilder $container)
         //mailing
         $transport = new Definition();
-        $transport->setClass('Swift_Transport');
+        $transport->setClass('TransportInterface');
```
`TransportInterface::class` (with import for ` Symfony\Component\Mailer\Transport\TransportInterface`)

Zowac replied:
Oops, fixed, thx :)

chalasr commented on main/core/Library/Mailing/Client/SymfonyMailer.php:
```diff
@@ -0,0 +1,46 @@
+
+            return true;
+        } catch (ErrorInterface $e) {
+            $this->logger->error($e->getMessage());
```
```suggestion
            $this->logger->error($e->getMessage(), ['exception' => $e]);
```

Zowac replied:
Done, thx :)

chalasr commented on main/core/Library/Mailing/TransportFactory.php:
```diff
@@ -12,72 +12,73 @@
+        $transport->setPassword($this->configHandler->getParameter('mailer_password'));
+        // should probably be configurable too
+        $transport->getStream()->setTimeout(self::TIMEOUT);
+        $transport->setSourceIp(null);
```
Does this method actually exist?

Zowac replied:
Yes, but i forget `getStream` before and after cheking the method doesn't support null value, i removed this line. Thanks you

chalasr commented on main/core/Library/Mailing/Client/SymfonyMailer.php:
```diff
@@ -0,0 +1,56 @@
+            $this->mailer->send($message);
+
+            return true;
+        } catch (ExceptionInterface $e) {
```
I just realized that the `Mailer` class is responsible for catching mailing-related errors and logging them to a specific file. Let's remove the try-catch

Zowac replied:
Indeed, thank you :)

Zowac committed:
fix transport

chalasr reviewed (approved):
👍  Tested locally, all good (both smtp and gmail).
Thanks @Zowac

Zowac commented:
Thanks for your help @chalasr

Elorfin commented on main/core/Library/Mailing/TransportFactory.php:
```diff
@@ -12,72 +12,70 @@
-            new \Swift_Events_SimpleEventDispatcher()
+        $transport = new EsmtpTransport(
+            $this->configHandler->getParameter('mailer_host'),
+            $this->configHandler->getParameter('mailer_port'),
```
It is `null` when using postal. It breaks the container compilation so we can't migrate it and it should be handled by code.

```
$this->configHandler->getParameter('mailer_port') ?? 0,
```

Elorfin commented on main/core/Library/Mailing/TransportFactory.php:
```diff
@@ -12,72 +12,70 @@
+            $this->logger
         );
+        $transport->setUsername($this->configHandler->getParameter('mailer_username'));
+        $transport->setPassword($this->configHandler->getParameter('mailer_password'));
```
Same thing with `mailer_username` and `mailer_transport`, `null` values are not allowed by `EsmtpTransport`.

Elorfin reviewed (changes requested):
Thank you for the work.
It just needs little changes to make `postal` work on our existing platforms.

Zowac committed:
Fix: Support for null values when using postal

Zowac commented:
@Elorfin 

The changes are done, thank you for your feedback :)
